back-end/src/sciler/communication/communication.go

```diff
@@ -0,0 +1,77 @@
+package communication
+
+import (
+	"errors"
+	"fmt"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	logger "github.com/sirupsen/logrus"
+	"time"
+)
+
+// Communicator is a type that maintains communication with the front-end and the client computers.
+type Communicator struct {
+	client           mqtt.Client
+	topicsOfInterest []string
+}
+
+// NewCommunicator is a constructor for a Communicator
+func NewCommunicator(host string, port string, topicsOfInterest []string) *Communicator {
+	opts := mqtt.NewClientOptions()
+	opts.AddBroker(fmt.Sprintf("%s://%s:%s", "tcp", host, port))
+	opts.SetClientID("back-end")
+	opts.SetConnectionLostHandler(onConnectionLost)
+	client := mqtt.NewClient(opts)
+	return &Communicator{client, topicsOfInterest}
+}
+
+// Start is a function that will start the communication by connecting to the broker and subscribing to all topics of interest
+func (communicator *Communicator) Start(handler mqtt.MessageHandler) {
+	_ = action(communicator.client.Connect, "connect", -1)
+	topics := make(map[string]byte)
+	for _, topic := range communicator.topicsOfInterest {
+		topics[topic] = byte(0)
+	}
+	_ = action(func() mqtt.Token {
+		return communicator.client.SubscribeMultiple(topics, handler)
+	}, "subscribing", -1)
+}
+
+func onConnectionLost(client mqtt.Client, e error) {
+	logger.Warn(fmt.Sprintf("Connection lost : %v", e))
+	if client.IsConnected() {
+		client.Disconnect(500)
+	}
+}
+
+// Publish is a method that will send a message to a specific topic
+func (communicator *Communicator) Publish(topic string, message string, retrials int) {
+	err := action(func() mqtt.Token {
+		return communicator.client.Publish(topic, byte(0), false, message)
+	}, "publish", retrials)
+	if err == errors.New("action failed") {
+		// todo reconnect?
```

Comment:
should this todo be solved? else make an issue probably

Reply:
created issue #49